### PR TITLE
IPv6/multi: Add length and type checks in ARP packet processing

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR adds checks for length and type (present in the hardware and protocol field) of an ARP header. These checks would stop ARP packets with incorrect values from getting through - such packets will be dropped.

Ref RFC 826.

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS_ARP.c [hunk 1]
```diff
@@ -135,83 +135,93 @@ eFrameProcessingResult_t eARPProcessPacket( NetworkBufferDescriptor_t * pxNetwor
...
+                {
+                    xARPHadIPClash = pdTRUE;
+                    /* Remember the MAC-address of the other device which has the same IP-address. */
+                    ( void ) memcpy( xARPClashMacAddress.ucBytes, pxARPHeader->xSenderHardwareAddress.ucBytes, sizeof( xARPClashMacAddress.ucBytes ) );
+                }
             }
-        }
-    #endif /* ipconfigARP_USE_CLASH_DETECTION */
+        #endif /* ipconfigARP_USE_CLASH_DETECTION */
+    }
```
Comment (htibosch) [inline FreeRTOS_ARP.c:224]: As I wrote in Chime, could you add a trace macro here, something like:
~~~c
if( good )
{
    /* Handle the packet. */
}
else
{
    iptraceDROPPED_INVALID_PACKET( pxARPHeader );
}
~~~

File: FreeRTOS_ARP.c [hunk 2]
```diff
@@ -135,83 +135,93 @@ eFrameProcessingResult_t eARPProcessPacket( NetworkBufferDescriptor_t * pxNetwor
...
-    /* The field ulSenderProtocolAddress is badly aligned, copy byte-by-byte. */
+    /* Only Ethernet hardware type is supported.
+     * Only IPv4 address can be present in the ARP packet.
+     * The hardware length (the MAC address) must be 6 bytes. And,
+     * The Protocol address length must be 4 bytes as it is IPv4. */
+    if( ( pxARPHeader->usHardwareType == ipARP_HARDWARE_TYPE_ETHERNET ) &&
+        ( pxARPHeader->usProtocolType == ipARP_PROTOCOL_TYPE ) &&
+        ( pxARPHeader->ucHardwareAddressLength == ipMAC_ADDRESS_LENGTH_BYTES ) &&
+        ( pxARPHeader->ucProtocolAddressLength == ipIP_ADDRESS_LENGTH_BYTES ) )
+    {
```
Comment (htibosch) [inline FreeRTOS_ARP.c:146]: I wrote:
~~~c
else
{
    iptraceDROPPED_INVALID_PACKET( pxARPHeader );
}
~~~
which should be `iptraceDROPPED_INVALID_ARP_PACKET()`.

**Review by htibosch — APPROVED** [review]
I approve this PR when the `iptrace` macro has been added.
Thanks again @AniruddhaKanhere

File: FreeRTOS_ARP.c (hunk 2 above)
Comment (AniruddhaKanhere) [reply FreeRTOS_ARP.c:146]: Done, thank you.

File: FreeRTOS_ARP.c (hunk 1 above)
Comment (AniruddhaKanhere) [reply FreeRTOS_ARP.c:224]: Yes, I added the trace macro. Thank you.

**Review by Bujain — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Pull request merged**